aten/src/ATen/native/mps/operations/LinearAlgebra.mm

```diff
@@ -91,6 +94,114 @@ void prepare_matrices_for_broadcasting(const Tensor* bias,
 
 enum LinearAlgebraOpType { ADDBMM_OP_TYPE, BADDBMM_OP_TYPE };
 
+void linalg_lu_factor_out_mps_impl(const Tensor& A, bool pivot, Tensor& LU, Tensor& pivots) {
+  using namespace mps;
+
+  Tensor A_t = A;
+  uint64_t aRows = A_t.size(-2);
+  uint64_t aCols = A_t.size(-1);
+  uint64_t aElemSize = A_t.element_size();
+  uint64_t numPivots = std::min(aRows, aCols);
+  resize_output(LU, A_t.sizes());
+  std::vector<int64_t> pivot_sizes(A_t.sizes().begin(), A_t.sizes().end() - 2);
+  pivot_sizes.push_back(numPivots);
+  resize_output(pivots, pivot_sizes);
+
+  if (A_t.numel() == 0 || LU.numel() == 0) {
+    LU.zero_();
+    return;
+  }
+
+  Tensor A_ = A_t.dim() > 3 ? A_t.flatten(0, -3) : A_t;
+  if (!A_.is_contiguous()) {
+    A_ = A_.clone(at::MemoryFormat::Contiguous);
+  }
+
+  uint64_t batchSize = A_.dim() > 2 ? A_.size(0) : 1;
+
+  std::vector<Tensor> status_tensors;
+  std::vector<Tensor> pivots_list;
+
+  status_tensors.reserve(batchSize);
+  pivots_list.reserve(batchSize);
+  for (C10_UNUSED const auto i : c10::irange(batchSize)) {
+    status_tensors.push_back(at::zeros(1, kInt, c10::nullopt, kMPS, c10::nullopt));
+    pivots_list.push_back(at::zeros(numPivots, kInt, c10::nullopt, kMPS, c10::nullopt));
+  }
```

Comment:
Why not just a tensor with `batchSize` elements in the case of `status_tensors`? Same same for `pivots_list`.

Reply:
I do not know enough about MPS, but perhaps @kulinseth can comment on what's the best way to do this. It'd be good if `pivots` could be a `Tensor`, because then we could implement the `_ex` version of the function, which would be the preferred way of implementing all this.

Reply:
Sorry, should it be initialized with `[ndArray buffer]` or `[arrayView buffer]`? I guess the latter is what you were suggesting.

With pivot matrix initialized with `[arrayView buffer]` that uses `MPSAliasingStrategyShallNotAlias`, the pivots output remains the same as the pivots before LU decomposition, which looks like the array view is not writable with this strategy? On the other hand, if I specify `MPSAliasingStrategyShallAlias`, the output is correct with unbatched inputs.

I also tried initializing with `[ndArray buffer]`, the outputs were incorrect if the inputs were batched.

Reply:
@kulinseth I can provide the code if you need repro.